EKF/control.cpp

```diff
@@ -757,10 +757,19 @@ void Ekf::controlHeightFusion()
 		// we have just switched to using range finder, calculate height sensor offset such that current
 		// measurment matches our current height estimate
 		if (_control_status_prev.flags.rng_hgt != _control_status.flags.rng_hgt) {
-			if (_terrain_initialised) {
-				_hgt_sensor_offset = _terrain_vpos;
+			// use the parameter rng_gnd_clearance if on ground to avoid a noisy offset initialization (e.g. sonar)
+			if (_control_status.flags.in_air) {
+
+				if (_terrain_initialised) {
+
+					_hgt_sensor_offset = _terrain_vpos;
+				} else {
+
+					_hgt_sensor_offset = _R_rng_to_earth_2_2 * _range_sample_delayed.rng + _state.pos(2);
```

Comment:
Yes, that is correct. My formulation was probably not very good. I just meant it behaves the same way as if you would have started from that location where you switched. 
And in most cases your suggestion would behave the exact same way. The only difference is that your suggestion would react with a sudden increase in height if the terrain is at that point not valid and in air. And before you even agreed that this is an unwanted behavior. So i'm wondering what you want to change. Do you think there should be an z estimate reset (e.g. to 0.5m in the above example)?